strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,106 @@ def context(self, *args, copies=1, shift="default"):
+            program_gates = [cmd.op.__class__.__name__ for cmd in self.rolled_circuit]
+            device_gates = [op["op"] for op in device_layout.operations]
+            if device_gates != program_gates:
+                raise CircuitError(
+                    "The gates or the order of gates used in the Program is incompatible with the device '{}' ".format(
+                        device.target
+                    )
+                )
+
+            # Second check: the gates act on the correct modes
+            program_modes = [[r.ind for r in cmd.reg] for cmd in self.rolled_circuit]
+            device_modes = [op["modes"] for op in device_layout.operations]
+            if program_modes != device_modes:
+                raise CircuitError(
+                    "Program cannot be used with the device '{}' "
+                    "due to incompatible mode ordering.".format(device.target)
+                )
+
+            # Third check: the parameters of the gates are valid
+            gate_params_ranges = device.gate_parameters
+            # We will loop over the different operations in the device specification
+
+            for i, operation in enumerate(device_layout.operations):
+                # We obtain the name of the parameter(s)
+                param_names = operation["args"]
+                program_params_len = len(self.rolled_circuit[i].op.p)
+                device_params_len = len(param_names)
+                # The next if is to make sure we do not flag incorrectly things like Sgate(r,0) being different Sgate(r)
+                # This assumes that a potential second parameter, that is not explicitly defined in the layout, is always 0.
+                if device_params_len < program_params_len:
+                    for j in range(1, program_params_len):
+                        if self.rolled_circuit[i].op.p[j] != 0:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter.".format(device.target)
+                            )
+                # Now we will check explicitly if the parameters in the program match
+                num_symbolic_param = 0  # counts the number of symbolic variables, which are labeled consecutively by the context method
+
+                for k, param_name in enumerate(param_names):
+                    # Obtain the relevant parameter range from the device
+                    param_range = device.gate_parameters[param_name]
+                    # Obtain the value of the corresponding parameter in the program
+                    program_param = self.rolled_circuit[i].op.p[k]
+
+                    if sf.parameters.par_is_symbolic(program_param):
+                        # If it is a symbolic value go and lookup its corresponding list in self.tdm_params
+                        self.assert_number_of_modes(device)
```

Comment:
Why is this only called if it's a symbolic parameter? Would this be best called simply at the top of this method, right under `if compiler == "td2" and device is not None:`, since we'd always like to check whether the correct number of modes are used? :thinking:

Reply:
You are completely right. There was direct logic testing for the number of temporal modes there. After you told about `assert_number_of_modes` I simply switched it, but it makes more sense to do the check right at the beginning. We  might have to change the test a bit since some warning will be triggered before other but that's OK.